{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "shared-renaissance",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%pylab inline\n",
    "\n",
    "# My implementation of method-of-moments\n",
    "\n",
    "# The program reads \"ps2_ibd.lwk.chr1.ped\", which contains chromosome 1 SNPs data\n",
    "# for 97 individuals, and estimates % IBD for each pair of individuals.\n",
    "\n",
    "# Outputs the result to \"my.ibd.genome\"\n",
    "\n",
    "import pandas as pd\n",
    "from itertools import combinations\n",
    "\n",
    "def main():\n",
    "    df = pd.read_csv(\"ps2_ibd.lwk.chr1.ped\", sep=\" \", header=None)\n",
    "    samples_to_alleles = {}\n",
    "    \n",
    "    for _, row in df.iterrows():\n",
    "        samples_to_alleles[row[0]] = list(row[6:])\n",
    "        \n",
    "    fout = open(\"my.ibd.genome\", 'w')\n",
    "    fout.write(\"ID1 ID2 Z0 Z1 Z2 PI\\n\") # header\n",
    "    \n",
    "    for id1, id2 in combinations(samples_to_alleles.keys(), 2): # for each pair\n",
    "        compute_Z(id1, id2, samples_to_alleles, fout)\n",
    "\n",
    "    fout.close()\n",
    "    \n",
    "\n",
    "def compute_Z(id1, id2, samples_to_alleles, fout):    \n",
    "    alleles1 = samples_to_alleles[id1]\n",
    "    alleles2 = samples_to_alleles[id2]\n",
    "    \n",
    "    N_IBS0, N_IBS1, N_IBS2 = 0, 0, 0\n",
    "       \n",
    "    N_I0Z0, N_I1Z0, N_I1Z1, N_I2Z0, N_I2Z1 = 0, 0, 0, 0, 0\n",
    "    N_I2Z2 = len(alleles1) / 2\n",
    "    \n",
    "    T_A = len(samples_to_alleles) * 2\n",
    "    T_A123 = (T_A / (T_A - 1)) * (T_A / (T_A - 2)) * (T_A / (T_A - 3))\n",
    "    T_A12 = (T_A / (T_A - 1)) * (T_A / (T_A - 2))\n",
    "            \n",
    "    for i in range(0, len(alleles1), 2): # for each SNP\n",
    "        X = 0 # allele 1 count\n",
    "        Y = 0 # allele 2 count\n",
    "        \n",
    "        genotype1 = alleles1[i] + alleles1[i+1]\n",
    "        genotype2 = alleles2[i] + alleles2[i+1]\n",
    "        \n",
    "        if abs(genotype1 - genotype2) == 2:\n",
    "            N_IBS0 += 1\n",
    "        elif abs(genotype1 - genotype2) == 1:\n",
    "            N_IBS1 += 1\n",
    "        else:\n",
    "            N_IBS2 += 1\n",
    "        \n",
    "        for alleles in samples_to_alleles.values():\n",
    "            for allele in (alleles[i], alleles[i+1]): \n",
    "                if allele == 1:\n",
    "                    X += 1\n",
    "                else:\n",
    "                    Y += 1\n",
    "                \n",
    "        p = X / T_A\n",
    "        q = Y / T_A\n",
    "        \n",
    "        if p != 0 and q != 0:\n",
    "            N_I0Z0 += 2 * pow(p, 2) * pow(q, 2) * ((X - 1) / X) * ((Y - 1) / Y) * T_A123\n",
    "\n",
    "            N_I1Z0 += 4 * pow(p, 3) * q * ((X - 1) / X) * ((X - 2) / X) * T_A123 + \\\n",
    "                      4 * p * pow(q, 3) * ((Y - 1) / Y) * ((Y - 2) / Y) * T_A123\n",
    "\n",
    "            N_I1Z1 += 2 * pow(p, 2) * q * ((X - 1) / X) * T_A12 + \\\n",
    "                      2 * p * pow(q, 2) * ((Y - 1) / Y) * T_A12\n",
    "        \n",
    "        if p == 0 and q != 0: \n",
    "            N_I2Z0 += pow(q, 4) * ((Y - 1) / Y) * ((Y - 2) / Y) * ((Y - 3) / Y) * T_A123\n",
    "            \n",
    "            N_I2Z1 += pow(q, 3) * ((Y - 1) / Y) * ((Y - 2) / Y) * T_A12\n",
    "        elif q == 0 and p != 0:\n",
    "            N_I2Z0 += pow(p, 4) * ((X - 1) / X) * ((X - 2) / X) * ((X - 3) / X) * T_A123\n",
    "            \n",
    "            N_I2Z1 += pow(p, 3) * ((X - 1) / X) * ((X - 2) / X) * T_A12  \n",
    "        elif p != 0 and q != 0:\n",
    "            N_I2Z0 += pow(p, 4) * ((X - 1) / X) * ((X - 2) / X) * ((X - 3) / X) * T_A123 + \\\n",
    "                      pow(q, 4) * ((Y - 1) / Y) * ((Y - 2) / Y) * ((Y - 3) / Y) * T_A123 + \\\n",
    "                      4 * pow(p, 2) * pow(q, 2) * ((X - 1) / X) * ((Y - 1) / Y) * T_A123\n",
    "\n",
    "            N_I2Z1 += pow(p, 3) * ((X - 1) / X) * ((X - 2) / X) * T_A12 + \\\n",
    "                      pow(q, 3) * ((Y - 1) / Y) * ((Y - 2) / Y) * T_A12 + \\\n",
    "                      pow(p, 2) * q * ((X - 1) / X) * T_A12 + \\\n",
    "                      p * pow(q, 2) * ((Y - 1) / Y) * T_A12\n",
    "        \n",
    "    Z0 = N_IBS0 / N_I0Z0\n",
    "    Z1 = (N_IBS1 - Z0 * N_I1Z0) / N_I1Z1\n",
    "    Z2 = (N_IBS2 - Z0 * N_I2Z0 - Z1 * N_I2Z1) / N_I2Z2\n",
    "    pi = Z1 / 2 + Z2\n",
    "    \n",
    "    Z0, Z1, Z2, pi = bound_Z(Z0, Z1, Z2, pi)\n",
    "    \n",
    "    fout.write(f\"{id1} {id2} {Z0} {Z1} {Z2} {pi}\\n\")\n",
    "\n",
    "def bound_Z(Z0, Z1, Z2, pi):\n",
    "    if Z0 > 1:\n",
    "        Z0, Z1, Z2 = 1, 0, 0\n",
    "    elif Z1 > 1:\n",
    "        Z0, Z1, Z2 = 0, 1, 0\n",
    "    elif Z2 > 1:\n",
    "        Z0, Z1, Z2 = 0, 0, 1   \n",
    "    elif Z0 < 0:\n",
    "        S = Z1 + Z2\n",
    "        Z0, Z1, Z2 = 0, Z1 / S, Z2 / S\n",
    "    elif Z1 < 0:\n",
    "        S = Z0 + Z2\n",
    "        Z0, Z1, Z2 = Z0 / S, 0, Z2 / S\n",
    "    elif Z2 < 0:\n",
    "        S = Z0 + Z1\n",
    "        Z0, Z1, Z2 = Z0 / S, Z1 / S, 0    \n",
    "    elif pow(pi, 2) <= Z2:\n",
    "        Z0 = (1 - pi) ** 2\n",
    "        Z1 = 2 * pi * (1 - pi)\n",
    "        Z2 = pi ** 2\n",
    "\n",
    "    return Z0, Z1, Z2, Z1 / 2 + Z2\n",
    "    \n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "breathing-pleasure",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# Run PLINK's method-of-moments on the same dataset\n",
    "# but with all autosomal SNPs\n",
    "\n",
    "# Outputs the result to \"plink.ibd.genome\"\n",
    "\n",
    "# DATADIR=/datasets/cs284-sp21-A00-public/ps2\n",
    "# plink --genome --bfile ${DATADIR}/ps2_ibd.lwk --out plink.ibd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "actual-chancellor",
   "metadata": {},
   "outputs": [],
   "source": [
    "%pylab inline\n",
    "\n",
    "# Plot IBD=0 vs IBD=1 for my result and PLINK's result\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "true_ibd = open(\"plink.ibd.genome\") # PLINK's result\n",
    "next(true_ibd)  # skip header\n",
    "\n",
    "IBD0, IBD1 = [], []\n",
    "\n",
    "for line in true_ibd.readlines():\n",
    "    row = line.split()\n",
    "    IBD0.append(float(row[6]))\n",
    "    IBD1.append(float(row[7]))\n",
    "\n",
    "plt.figure()\n",
    "plt.scatter(IBD0, IBD1, s=15)\n",
    "plt.title(\"PLINK's Implementation\")\n",
    "plt.xlabel(\"IBD0\")\n",
    "plt.ylabel(\"IBD1\")\n",
    "plt.show()\n",
    "\n",
    "ibd = open(\"my.ibd.genome\") # my result\n",
    "next(ibd)  # skip header\n",
    "\n",
    "IBD0, IBD1 = [], []\n",
    "\n",
    "for line in ibd.readlines():\n",
    "    row = line.split()\n",
    "    IBD0.append(float(row[2]))\n",
    "    IBD1.append(float(row[3]))\n",
    "\n",
    "plt.figure()\n",
    "plt.scatter(IBD0, IBD1, s=15)\n",
    "plt.title(\"My Implementation\")\n",
    "plt.xlabel(\"IBD0\")\n",
    "plt.ylabel(\"IBD1\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "animated-commons",
   "metadata": {},
   "outputs": [],
   "source": [
    "%pylab inline\n",
    "\n",
    "# Calculate MSE for pi, which is P(IBD = 2) + 0.5 * P(IBD = 1),\n",
    "# between my result and PLINK's result\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "plink_ibd = open(\"plink.ibd.genome\")\n",
    "my_ibd = open(\"my.ibd.genome\")\n",
    "next(plink_ibd)\n",
    "next(my_ibd)\n",
    "\n",
    "plink_pi, my_pi = [], []\n",
    "\n",
    "for line in plink_ibd.readlines():\n",
    "    row = line.split()\n",
    "    plink_pi.append(float(row[9]))\n",
    "\n",
    "for line in my_ibd.readlines():\n",
    "    row = line.split()\n",
    "    my_pi.append(float(row[5]))\n",
    "\n",
    "MSE = sum(pow((np.array(plink_pi) - np.array(my_pi)), 2)) / len(plink_pi)\n",
    "print(MSE)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
